File: section12#Math_Set/C++/mathSet.cpp
#include <iostream>
#include <string>
#include <opencv2/opencv.hpp>


#define BONE_PIC	"Fig0232(a)(partial_body_scan).tif"

using namespace std;
using namespace cv;

static bool complementSetTest(Mat imgOri);

int main()
{
	Mat imgBoneOri = imread(BONE_PIC);
	if (imgBoneOri.empty())
	{
		cout << "Cannot find this picture!" << endl;
		return false;
	}
	imshow("boneOri", imgBoneOri);


	complementSetTest(imgBoneOri);

	waitKey(0);
	return true;
}

static bool complementSetTest(Mat imgOri)
{
	Mat imgOut(imgOri.size(), imgOri.type(), Scalar(255, 255, 255));
	if (CV_8UC3 != imgOri.type())
	{
		cout << "Can not match this picture type!" << endl;
		return false;
	}

	//imgOut = imgOut - imgOri;

	bitwise_not(imgOri, imgOut);

	imshow("imgBoneCom", imgOut);
	imwrite("BoneCom.png", imgOut);

	return true;
}

File: section12#Math_Set/Python/mathSet.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# author：lowkeyway time:7/19/2019

import sys
import cv2 as cv
import numpy as np

def main_func(argv):
    imgBone = cv.imread("Fig0232(a)(partial_body_scan).tif")
    cv.imshow("boneOri", imgBone)

    imgOut = np.zeros(imgBone.shape, imgBone.dtype)
    cv.bitwise_not(imgBone, imgOut)
    cv.imshow("boneOut", imgOut)
    cv.imwrite("bonBitwiseNot.png", imgOut)


    cv.waitKey(0)

if __name__ == '__main__':
    main_func(sys.argv)